File: src/main/java/com/flexpoker/processmanagers/IncrementBlindsCountdownProcessManager.java
package com.flexpoker.processmanagers;

import com.flexpoker.framework.command.CommandSender;
import com.flexpoker.framework.processmanager.ProcessManager;
import com.flexpoker.game.command.commands.GameCommand;
import com.flexpoker.game.command.commands.IncrementBlindsCommand;
import com.flexpoker.game.command.events.GameStartedEvent;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import javax.inject.Inject;
import java.util.Timer;
import java.util.TimerTask;

@Component
public class IncrementBlindsCountdownProcessManager implements ProcessManager<GameStartedEvent> {

    private final CommandSender<GameCommand> gameCommandSender;

    @Inject
    public IncrementBlindsCountdownProcessManager(CommandSender<GameCommand> gameCommandSender) {
        this.gameCommandSender = gameCommandSender;
    }

    @Async
    @Override
    public void handle(GameStartedEvent event) {
        addNewBlindIncrementTimer(event);
    }

    private void addNewBlindIncrementTimer(GameStartedEvent event) {
        var actionOnTimer = new Timer();
        var timerTask = new TimerTask() {
            @Override
            public void run() {
                var command = new IncrementBlindsCommand(event.getAggregateId());
                gameCommandSender.send(command);
            }
        };

        var blindIncrementInMilliseconds = event.getBlindScheduleDTO().getNumberOfMinutesBetweenLevels() * 60000;
        actionOnTimer.scheduleAtFixedRate(timerTask, blindIncrementInMilliseconds, blindIncrementInMilliseconds);
    }

}

File: src/main/kotlin/com/flexpoker/processmanagers/IncrementBlindsCountdownProcessManager.kt
package com.flexpoker.processmanagers

import com.flexpoker.framework.command.CommandSender
import com.flexpoker.framework.processmanager.ProcessManager
import com.flexpoker.game.command.commands.GameCommand
import com.flexpoker.game.command.commands.IncrementBlindsCommand
import com.flexpoker.game.command.events.GameStartedEvent
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.atomic.AtomicInteger
import javax.inject.Inject

@Component
class IncrementBlindsCountdownProcessManager @Inject constructor(
    private val gameCommandSender: CommandSender<GameCommand>
) : ProcessManager<GameStartedEvent> {

    @Async
    override fun handle(event: GameStartedEvent) {
        val actionOnTimer = Timer()
        val timerTask: TimerTask = object : TimerTask() {
            val numberOfTimesToRun = event.blindSchedule.maxLevel() - 1
            val numberOfExecutions: AtomicInteger = AtomicInteger(0)

            override fun run() {
                val command = IncrementBlindsCommand(event.aggregateId)
                gameCommandSender.send(command)

                if (numberOfExecutions.incrementAndGet() == numberOfTimesToRun) {
                    actionOnTimer.cancel()
                }
            }
        }
        val blindIncrementInMilliseconds = event.blindSchedule.numberOfMinutesBetweenLevels * 60000
        actionOnTimer.scheduleAtFixedRate(
            timerTask,
            blindIncrementInMilliseconds.toLong(),
            blindIncrementInMilliseconds.toLong()
        )
    }

}